src/main.rs

```diff
@@ -34,33 +35,51 @@ async fn run_broadcast(cli: &Config) -> Result<(), openssh::Error> {
     let hosts = get_hosts();
     let num_hosts = hosts.len();
 
+    // Set handler for Ctrl-c. This will set the `cancelled` variable to
+    // true, which will be noticed by the scheduling loop.
+    let cancelled = Arc::new(Mutex::new(false));
+    let cancelled_handler = Arc::clone(&cancelled);
+    ctrlc::set_handler(move || {
+        eprintln!("\n[Pegasus] Ctrl-c detected. Sending out cancel notices.");
+        *cancelled_handler.blocking_lock() = true;
+    })
+    .expect("Failed to set ctrl-c handler.");
+
     // Broadcast channel used to distribute commands to all hosts.
     let (command_tx, _) = broadcast::channel::<Cmd>(1);
 
     // Synchronization barrier that all SSH sessions report to when
     // they're done executing. The scheduler also waits on this barrier
     // to know when all sessions are done.
-    let barrier = Arc::new(Barrier::new(num_hosts + 1));
+    let end_barrier = Arc::new(Barrier::new(num_hosts + 1));
+
+    // An atomic variable set whenever a session errors. Later read by
+    // the scheduling loop to determine whether or not to exit.
+    let errored = Arc::new(AtomicBool::new(false));
 
-    let mut tasks = vec![];
-    let errored = Arc::new(Mutex::new(false));
+    let mut tasks = Vec::with_capacity(num_hosts);
     let colors = ColorPalette::new(num_hosts as u32, PaletteType::Pastel, false).colors;
     for (color, host) in zip(colors, hosts) {
         let mut command_rx = command_tx.subscribe();
-        let barrier = Arc::clone(&barrier);
+        let end_barrier = Arc::clone(&end_barrier);
         let errored = Arc::clone(&errored);
         tasks.push(tokio::spawn(async move {
             // Open a new SSH session with the host.
             let session = Session::connect(host, color).await;
             // Handlebars registry for filling in parameters.
             let mut registry = Handlebars::new();
-            while let Ok(job) = command_rx.recv().await {
-                let job = job.fill_template(&mut registry, &session.host);
-                let result = session.run(job).await;
+            // When cancellation is triggered by the ctrlc handler, the
+            // scheduling loop will see that, break, and drop `command_tx`.
+            // Then `command_rx.recv()` will return `Err`, allowing the
+            // session object to be dropped, and everyting gracefully
+            // terminated.
+            while let Ok(cmd) = command_rx.recv().await {
+                let cmd = cmd.fill_template(&mut registry, &session.host);
+                let result = session.run(cmd).await;
                 if result.is_err() || result.unwrap().code() != Some(0) {
-                    *errored.lock().await = true;
+                    errored.store(true, Ordering::Relaxed);
```

Comment:
Why not return the error directly and use `task.await` to retrieve the error?

Reply:
`task` is currently persistent, where it internally opens a single SSH session with its assigned host and receives commands from channels to execute it. (Basically it shares its lifetime with the SSH session.) So the tasks aren't really designed to return anything. Also, since `session.run` prints out the error to the stdout of Pegasus, there's really no point for the scheduler in knowing about what kind of error that was.

Reply:
OK, got it.